### DISCUSSION TITLE
Support for waveshare e_paper with pico

### DISCUSSION THREAD

**Discussion by TangerineDreamer** [Bodmer/TFT_eSPI#2114] [New processor/display support or capability request]
Hi,bodmer,
It's pretty hard to find something working for the e_paper shield that hold a pico,only a difficult py or c programmation examples,
or also for arduino boards .The tft e_spi propose the steinberg example,but only for esp8266,with pre-defined pin settings.
The settings used for pico e_paper 2.13 are:

https://www.waveshare.com/wiki/Pico-ePaper-2.13

DIN | GP11 | MOSI pin of SPI interface, data transmitted from Master to Slave.
CLK | GP10 | SCK pin of SPI interface, clock input
CS | GP9 | Chip select pin of SPI interface, Low active
DC | GP8 | Data/Command control pin (High: Data; Low: Command)
RST | GP12 | Reset pin, low active
BUSY | GP13 | Busy pin

How can i adapt those settings to e_spi?
Or will there be a future support?

**Comment by Bodmer** [comment]
The ePaper support was added before the RP2040 even existed and as you note is for the ESP8266 only.

The RP2040 is my favorite processor so I will have a look at getting the example and support EDP library to work with it. I think this should be quite easy... hopefully! I have an ePaper 2.13 display somewhere that I can test it with, finding that might be more difficult...

**Comment by Bodmer** [comment]
Here is a zip file for the library for your display. Load it as an Arduino library. Try it and tell me if it works. There is an example in that library.
[epd2in13.zip](https://github.com/Bodmer/TFT_eSPI/files/9874443/epd2in13.zip)

Note that your GPIO choice above is hard coded into that library in the epdif.h file.

You will need to define a user setup file for TFT_eSPI. Here is an example content:
```
#define EPD_DRIVER // ePaper driver

#define TFT_MISO       -1
#define TFT_MOSI       -1
#define TFT_SCLK       -1
#define TFT_RST        -1

#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:-.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts

#define SMOOTH_FONT
```

The Floyd_Steinberg TFT_eSPI library example should now run. Ignore the fact that is it appears to use SPIFFS, LittleFS will in fact be used automatically. To get the tiger image included in the sketch data folder loaded into the pico define a FS partition in the IDE using the menu option, say sketch 1MB and FS 1MB, and upload the data from the sketch using the IDE LittleFS data upload menu option. Then compile and upload the sketch. Hopefully it will run but the tiger image will be bigger than the display so will only partially show.

**Reply by Bodmer** [reply]
You will need to comment out this line:
https://github.com/Bodmer/TFT_eSPI/blob/master/examples/ePaper/Floyd_Steinberg/Floyd_Steinberg.ino#L40

And uncomment this line:
https://github.com/Bodmer/TFT_eSPI/blob/master/examples/ePaper/Floyd_Steinberg/Floyd_Steinberg.ino#L41

To select the correct display library.

**Comment by TangerineDreamer** [comment]
Hello,
It's very kind of you for your help,
I've uploaded the data in littlefs
I've done all you said,took your "epd2in13.zip",tried the demo inside,then the steinberg example,all compiled and uploaded
perfectly,the procedure in the monitor shows:
...Waking up!
...Updating display
but the e_paper don't change at all!
Must be i've forgotten something...

**Reply by Bodmer** [reply]
Maybe I forgot something. I got my 2.9" display working with the Pico OK.  I can't find a 2.13" one so just copied the changes over for the smaller display. I will have another look for a smaller display.

**Reply by Bodmer** [reply]
Run Read_User_Setup to check the correct setup is being compiled in. Post the output.

**Comment by TangerineDreamer** [comment]
Here it is,but no mention about the pins used...

[code]
TFT_eSPI ver = 2.4.78
Processor    = RP2040
Transactions = Yes
Interface    = SPI
Display driver = ePaper


Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled

Display SPI frequency = 20.00
[/code]

**Reply by Bodmer** [reply]
The missing info is OK as the pins are defined in the ePaper support library. The output looks OK as ePaper is reported in the display type.

I see on the Waveshare site there are various versions of the display and these appear to need different command sequences. Do you know which ePaper display version you have? The driver I posted is circa 4 years old and perhaps the displays have been updated.

**Comment by TangerineDreamer** [comment]
This one:

https://fr.aliexpress.com/item/1005004560859817.html?spm=a2g0o.productlist.0.0.7c801c0aCC2R4C&algo_pvid=b5a44087-6cfe-4db2-a8bc-8a33d712a91a&algo_exp_id=b5a44087-6cfe-4db2-a8bc-8a33d712a91a-5&pdp_ext_f=%7B%22sku_id%22%3A%2212000029620411584%22%7D&pdp_npi=2%40dis%21EUR%2130.18%2122.33%21%21%21%21%21%400b0a0ae216669453406228561e7e40%2112000029620411584%21sea&curPageLogUid=QskwIqxqP1V3

there are some indications on the strap wire :  HINK-E021-3A22-A0   SLH 1948

I searched over the web with those indications and found somethink that could interest you:

https://devzone.nordicsemi.com/f/nordic-q-a/87352/waveshare-e-paper-with-ssd16xx-driver

**Reply by Bodmer** [reply]
That is a link to a 3 colour ePaper display. The first link posted is to a 2 colour (balack and white display. These are different types and thus need quite different drivers.

Plaease confirm which display you have as clearly one of these links is misleading.

**Comment by TangerineDreamer** [comment]
You're right it's a bit confusing,here is the exact link,but i can't find no more about my specifications order,

https://fr.aliexpress.com/item/1005002461877666.html?spm=a2g0o.order_list.0.0.21ef5e5b2vpxwC&gatewayAdapt=glo2fra

i've tried with a  2colors demo uf2,it's ok,but i can't find with 3 colors to try with,if you know where i can download them
i can verify if it's a 3 colors.

**Comment by TangerineDreamer** [comment]
I've made a total reset of the pico flash by holding RUN to GND then hold BOOTSEL,then release RUN from GND before
release BOOTSEL after the UF2 board appears.
Then i upload the 2.13 epd.uf2 and it appears this one makes the screen react as the demo do,with monitor debug:

Clear...
Goto Sleep...
close 5V, Module enters 0 power consumption ...
EPD_2IN13_V2_test Demo
DEV_Module_Init OK 
e-Paper Init and Clear...
Paint_NewImage
show image for array
Drawing
Drawing
Partial refresh

Seems only the 2.13 (extra_uf2) works,not the 2.13b or 2.13c or 2.13d,so it's "EPD_2IN13_V2_test" as the monitor shows,but without
other epd.uf2 to test,i can't tell you if the "EPD_2IN13_V3_test" will also work

[Pico_ePaper_Code.zip](https://github.com/Bodmer/TFT_eSPI/files/9896033/Pico_ePaper_Code.zip)
